scripts/python/airflow/dags/contralory_declaration_data_extraction.py

```diff
@@ -0,0 +1,194 @@
+from json import loads
+from json import dumps
+from time import sleep
+from requests import post
+from random import randint
+from string import Template
+from datetime import timedelta
+from airflow.hooks.postgres_hook import PostgresHook
+from airflow.models import DAG
+from airflow.models import Variable
+from airflow.operators.dummy_operator import DummyOperator
+from airflow.operators.postgres_operator import PostgresOperator
+from airflow.operators.python_operator import PythonOperator
+from airflow.utils.dates import days_ago
+
+dag_timer = int(Variable.get("CGR_PARSER_TIMER", 0))
+dag_services = str(Variable.get("CGR_PARSER_SERVICES", "http://localhost:8000/parser/send")).split(",")
+dag_year_filter = int(Variable.get("CGR_PARSER_MIN_YEAR", 2015))
+dag_sub_jobs_count = int(Variable.get("CGR_PARSER_SUB_JOBS_COUNT", 16))
+dag_sub_jobs_batch_size = int(Variable.get("CGR_PARSER_PDF_SUB_JOBS_BATCH_SIZE", 10))
+
+default_args = {
+	"owner": "airflow",
+	"depends_on_past": False,
+	"email": ["arturovolpe@gmail.com"],
+	"email_on_failure": False,
+	"email_on_retry": False,
+	"retry_delay": timedelta(hours=1)
+}
+
+# parser request template
+parser_request = {
+	"file": {
+		"path": "$resource"
+	}
+}
+
+dag = DAG(
+	dag_id="contralory_declaration_data_extraction",
+	default_args=default_args,
+	description="CGR data extraction with concurrent calls to ms_djbr_parser API",
+	start_date=days_ago(1),
+	schedule_interval=timedelta(weeks=1),
+)
+
+def is_valid_data(data):
+	if data is None:
+		return False
+
+	if not 'resumen' in data.keys():
+		return False
+
+	items = list(data['resumen'].values())
+
+	if len(items) != 3:
+		return False
+
+	if sum(map(abs, items)) == 0:
+		return False
+
+	return True
+
+# redirect the request traffic
+def get_random_url():
+	return dag_services[randint(0, len(dag_services) - 1)]
+
+# arithmetic progression generator
+def ap(pos, first_item, second_item):
+	diff = second_item - first_item
+	index = pos
+
+	# generare index item
+	def generate():
+		nonlocal diff
+		nonlocal index
+		index += 1
+		return first_item + ((index - 1) - 1) * diff
+
+	return generate
+
+def list_navigator(cursor, ap, year, batch_size):
+	should_continue = True
+
+	while should_continue:
+		cursor.execute("""
+		SELECT drd.id as raw_id, ddf.file_name from staging.djbr_downloaded_files ddf
+		INNER JOIN staging.djbr_raw_data drd on ddf.raw_data_id = drd.id
+		WHERE drd.periodo >= %s
```

Comment:
We can add here a JOIN to prevent re-parsing data:
```suggestion
                LEFT JOIN analysis.ddjj_readable_data parsed ON parsed.raw_data_id = drd.id
		WHERE drd.periodo >= %s AND parsed IS NULL
```